File: go/crc.go
package embedkv

import "hash/crc32"

var crcTable = crc32.MakeTable(crc32.IEEE)

// computeBlockCRC32 computes the CRC32 of buf, treating bytes [crcOffset, crcOffset+4)
// as 0x00000000. buf must be the full block (all blockSize bytes).
func computeBlockCRC32(buf []byte, crcOffset int) uint32 {
	h := crc32.New(crcTable)
	h.Write(buf[:crcOffset])
	h.Write([]byte{0, 0, 0, 0})
	h.Write(buf[crcOffset+4:])
	return h.Sum32()
}

// blockCRCOffset returns the offset of the block_crc32 field, which is always
// the last 4 bytes of the block regardless of block type.
func blockCRCOffset(buf []byte) int { return len(buf) - 4 }

// writeBlockCRC computes the CRC over the whole block (treating the trailing
// CRC field as zero) and stores it at block_size-4.
func writeBlockCRC(buf []byte) {
	off := blockCRCOffset(buf)
	le.PutUint32(buf[off:], computeBlockCRC32(buf, off))
}

// verifyBlockCRC returns true when the stored trailing CRC matches the block.
func verifyBlockCRC(buf []byte) bool {
	off := blockCRCOffset(buf)
	return computeBlockCRC32(buf, off) == le.Uint32(buf[off:])
}

File: rust/src/crc.rs
use std::sync::OnceLock;

static CRC_TABLE: OnceLock<[u32; 256]> = OnceLock::new();

fn crc_table() -> &'static [u32; 256] {
    CRC_TABLE.get_or_init(|| {
        let poly: u32 = 0xEDB8_8320; // IEEE reflected polynomial
        let mut table = [0u32; 256];
        for i in 0u32..256 {
            let mut crc = i;
            for _ in 0..8 {
                if crc & 1 != 0 {
                    crc = (crc >> 1) ^ poly;
                } else {
                    crc >>= 1;
                }
            }
            table[i as usize] = crc;
        }
        table
    })
}

fn crc32_update(mut crc: u32, data: &[u8]) -> u32 {
    let table = crc_table();
    for &b in data {
        let idx = ((crc ^ (b as u32)) & 0xFF) as usize;
        crc = (crc >> 8) ^ table[idx];
    }
    crc
}

/// Compute the CRC32 (IEEE) of `buf`, treating the 4 bytes at `crc_offset`
/// as 0x00000000. `buf` must be the full block.
pub fn compute_block_crc32(buf: &[u8], crc_offset: usize) -> u32 {
    let mut crc = 0xFFFF_FFFFu32;
    crc = crc32_update(crc, &buf[..crc_offset]);
    crc = crc32_update(crc, &[0u8; 4]);
    crc = crc32_update(crc, &buf[crc_offset + 4..]);
    crc ^ 0xFFFF_FFFF
}

/// Offset of the block_crc32 field: always the last 4 bytes of the block,
/// regardless of block type.
pub fn block_crc_offset(buf: &[u8]) -> usize {
    buf.len() - 4
}

/// Compute the block CRC (treating the trailing CRC field as zero) and store it
/// in the last 4 bytes of the block.
pub fn write_block_crc(buf: &mut [u8]) {
    let off = block_crc_offset(buf);
    let crc = compute_block_crc32(buf, off);
    buf[off..off + 4].copy_from_slice(&crc.to_le_bytes());
}

/// Return true when the stored trailing CRC matches the block contents.
pub fn verify_block_crc(buf: &[u8]) -> bool {
    let off = block_crc_offset(buf);
    let stored = u32::from_le_bytes([buf[off], buf[off + 1], buf[off + 2], buf[off + 3]]);
    compute_block_crc32(buf, off) == stored
}
